ext/drreg/drreg.c

```diff
@@ -1599,19 +2410,62 @@ drreg_restore_app_aflags(void *drcontext, instrlist_t *ilist, instr_t *where)
  */
 
 static bool
-is_our_spill_or_restore(void *drcontext, instr_t *instr, bool *spill OUT,
-                        reg_id_t *reg_spilled OUT, uint *slot_out OUT, uint *offs_out OUT)
+is_our_spill_or_restore(void *drcontext, instr_t *instr, instr_t *next_instr,
+                        bool *spill OUT, reg_id_t *reg_spilled OUT, uint *slot_out OUT,
+                        uint *offs_out OUT, bool *is_indirectly_spilled OUT)
 {
     bool tls;
     uint slot, offs;
     reg_id_t reg;
-    if (!instr_is_reg_spill_or_restore(drcontext, instr, &tls, spill, &reg, &offs))
+    bool is_spilled;  /* Flag denoting spill or restore*/
+    bool is_indirect; /* Flag denoting direct or indirect access */
+
+    is_indirect = false;
+    if (is_indirectly_spilled != NULL)
+        *is_indirectly_spilled = is_indirect;
+
+    if (!instr_is_reg_spill_or_restore(drcontext, instr, &tls, &is_spilled, &reg, &offs))
         return false;
-    /* Is this from our raw TLS? */
+    /* Is this from our direct raw TLS for gpr registers ? */
     if (tls && offs >= tls_slot_offs &&
         offs < (tls_slot_offs + ops.num_spill_slots * sizeof(reg_t))) {
         slot = (offs - tls_slot_offs) / sizeof(reg_t);
-    } else {
+    }
+#ifdef X86
+    else if (tls && offs == tls_simd_offs &&
+             !(is_spilled) /* Cant be a spill bc loading block */) {
+        /* In order to detect indirect spills, the loading of the pointer
+         * to the indrect block must be done exactly prior. We assume that
+         * nobody else can interfere with our indirect load sequence for
+         * simd registers.
+         */
+        ASSERT(next_instr != NULL, "next_instr cannot be NULL");
+        ASSERT(instr_get_opcode(next_instr) == OP_movdqa ||
+                   instr_get_opcode(next_instr) == OP_vmovdqa,
```

Comment:
When YMM and ZMM spills are supported, the developer might need to update the ASSERT here.

Reply:
Done